### PR TITLE
Complement QoL changes

### PR DESCRIPTION
**Pull request by S7evinK**
This PR does the following:
- adds a `keysize` parameter to `generate-keys`, so we can use lower sized keys when running in CI
- updates the Complement docker files to use BuildKit (requires Docker >18.09)
- uses `exec` when executing `dendrite-monotlith-server`, making it PID 1 inside docker, which results in Dendrite actually receiving the `SIGTERM` signal send by Docker. (Making it faster when running tests with Complement, as we don't take 10 seconds to timeout)

CC: @MadLittleMods I think you were complaining about longer build/execution times, iirc?

### CODE REVIEW

File: build/scripts/ComplementPostgres.Dockerfile [hunk 1]
```diff
@@ -45,10 +45,10 @@ EXPOSE 8008 8448
...
+CMD /build/run_postgres.sh && ./generate-keys --keysize 64 --server $SERVER_NAME --tls-cert server.crt --tls-key server.key --tls-authority-cert /complement/ca/ca.crt --tls-authority-key /complement/ca/ca.key && \
     ./generate-config -server $SERVER_NAME --ci > dendrite.yaml && \
     # Replace the connection string with a single postgres DB, using user/db = 'postgres' and no password, bump max_conns
     sed -i "s%connection_string:.*$%connection_string: postgresql://postgres@localhost/postgres?sslmode=disable%g" dendrite.yaml && \
     sed -i 's/max_open_conns:.*$/max_open_conns: 100/g' dendrite.yaml && \
     cp /complement/ca/ca.crt /usr/local/share/ca-certificates/ && update-ca-certificates && \
-    ./dendrite-monolith-server --really-enable-open-registration --tls-cert server.crt --tls-key server.key --config dendrite.yaml -api=${API:-0}
\ No newline at end of file
+    exec ./dendrite-monolith-server --really-enable-open-registration --tls-cert server.crt --tls-key server.key --config dendrite.yaml -api=${API:-0}
```
Comment (MadLittleMods) [inline build/scripts/ComplementPostgres.Dockerfile:54]: >CC: @MadLittleMods I think you were complaining about longer build/execution times, iirc?
 
 Yes! Especially with Synapse, https://github.com/matrix-org/synapse/issues/13204

**Review by neilalexander — APPROVED** [review]
LGTM, although might be best to run it by @kegsay too.

File: build/scripts/Complement.Dockerfile [hunk 2]
```diff
@@ -26,7 +26,7 @@ EXPOSE 8008 8448
 
 # At runtime, generate TLS cert based on the CA now mounted at /ca
 # At runtime, replace the SERVER_NAME with what we are told
-CMD ./generate-keys --server $SERVER_NAME --tls-cert server.crt --tls-key server.key --tls-authority-cert /complement/ca/ca.crt --tls-authority-key /complement/ca/ca.key && \
+CMD ./generate-keys -keysize 1024 --server $SERVER_NAME --tls-cert server.crt --tls-key server.key --tls-authority-cert /complement/ca/ca.crt --tls-authority-key /complement/ca/ca.key && \
     ./generate-config -server $SERVER_NAME --ci > dendrite.yaml && \
     cp /complement/ca/ca.crt /usr/local/share/ca-certificates/ && update-ca-certificates && \
-    ./dendrite-monolith-server --really-enable-open-registration --tls-cert server.crt --tls-key server.key --config dendrite.yaml -api=${API:-0}
+    exec ./dendrite-monolith-server --really-enable-open-registration --tls-cert server.crt --tls-key server.key --config dendrite.yaml -api=${API:-0}
```
Comment (kegsay) [inline build/scripts/Complement.Dockerfile:32]: What difference does this make in practice?

**Review by kegsay — APPROVED** [review]
It would be nice to see some actual timings before/after.

File: build/scripts/Complement.Dockerfile (hunk 2 above)
Comment (neilalexander) [reply build/scripts/Complement.Dockerfile:32]: Seeming 5m 49s on this PR, vs 7m 5s on the last `main` run.

Comment (S7evinK) [reply build/scripts/Complement.Dockerfile:32]: For Postgres it's almost 5m faster (8m 19s vs 12m 49s)

File: build/scripts/ComplementPostgres.Dockerfile (hunk 1 above)
Comment (S7evinK) [reply build/scripts/ComplementPostgres.Dockerfile:54]: Looks like Synapse is already PID 1 inside the container and receives the `SIGTERM` request, so this is not the issue. :/

**Pull request merged**